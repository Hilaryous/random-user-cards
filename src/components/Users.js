// @flow
import React, { PureComponent } from 'react'
import map from 'ramda/src/map'
import User from './User'
import Posts from './Posts'
import CardGrid from './CardGrid'
import type { UserType } from '../types/User'
import { getUsers } from '../utils'

type Props = {}

type State = {
  error: ?{
    message: '',
  },
  isLoaded: boolean,
  users: Array<UserType>,
  currentUserId: string,
}

class Users extends PureComponent<Props, State> {
  state = {
    isLoaded: false,
    users: [],
    currentUserId: '',
    error: null,
  };

  componentDidMount() {
    fetch('https://jsonplaceholder.typicode.com/users')
      .then(res => res.json())
      .then(
        (result) => {
          this.setState(() => ({
            isLoaded: true,
            users: getUsers(result, 3),
          }))
        },
        (error) => {
          this.setState(() => ({
            isLoaded: true,
            error,
          }))
        },
      )
  }

  toggleUser = (id: string) => {
    this.setState(({ currentUserId }) => ({
      currentUserId: currentUserId !== id ? id : '',
    }))
  }

  render() {
    const { currentUserId, error, isLoaded, users } = this.state
    if (error) {
      return (
        <div>
          Error:
          {error.message}
        </div>
      )
    } if (!isLoaded) {
      return (
        <div>
          Loading...
        </div>
      )
    }
    return (
      <CardGrid>
        { map(({ avatar, catchPhrase, email, id, name }) => (
          <User
            avatar={avatar}
            catchPhrase={catchPhrase}
            currentUserId={currentUserId}
            email={email}
            id={id}
            key={id}
            name={name}
            toggleUser={this.toggleUser}
          />
        ), users)}
        <Posts userId={currentUserId} />
      </CardGrid>
    )
  }
}

export default Users
