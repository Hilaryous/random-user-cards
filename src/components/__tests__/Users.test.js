/* eslint-disable function-paren-newline */
import React from 'react'
import renderer from 'react-test-renderer'
import Users from '../Users'

describe('Users', () => {
  it('shows the correct snapshot tree for the Users component, loading', () => {
    const tree = renderer.create(
      <Users />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
