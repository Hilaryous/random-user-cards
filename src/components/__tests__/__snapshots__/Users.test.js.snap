// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users shows the correct snapshot tree for the Users component, loading 1`] = `
<div>
  Loading...
</div>
`;
